subworkflows/local/genotype.nf

```diff
@@ -0,0 +1,373 @@
+                has_aux: true
+            }
+
+        // mix the two branches back together after fixing cardinality
+        ch_refs_for_mpileup_pileupcaller = ch_refs_prep.no_aux
+            .map {
+                ref_meta, fasta, fai, dict, empty ->
+                [ ref_meta, fasta, fai, dict, [], [] ]
+            }
+            .mix( ch_refs_prep.has_aux )
+            .map {
+            // Prepend a new meta that contains the meta.id value as the new_meta.reference attribute
+                WorkflowEager.addNewMetaFromAttributes( it, "id" , "reference" , false )
+            } // RESULT: [ [combination_meta], [ref_meta], fasta, fai, dict, bed, snp ]
+
+        // Prepare collect bams for mpileup
+        ch_mpileup_inputs_bams = ch_bam_bai
+            .map {
+                WorkflowEager.addNewMetaFromAttributes( it, ["reference", "strandedness"] , ["reference", "strandedness"] , false )
+            }
+            .groupTuple()
+            .map {
+                combo_meta, metas, bams, bais ->
+                def ids = metas.collect { meta -> meta.id }
+                [ combo_meta + [id: ids], bams ] // Drop bais
+            } // Collect all IDs into a list in meta.id. Useful when running pileupCaller later
+
+            // Combine prepped bams and references
+            ch_mpileup_inputs = ch_mpileup_inputs_bams
+                .map {
+                    WorkflowEager.addNewMetaFromAttributes( it, "reference", "reference" , false )
+                }
+                .combine( ch_refs_for_mpileup_pileupcaller , by:0 )
+                // do not run if no bed file is provided
+                .filter { it[7] != []}
+                .multiMap {
+                    ignore_me, combo_meta, bams, ref_meta, fasta, fai, dict, bed, snp ->
+                        def bedfile = bed != "" ? bed : []
+                        bams:  [ combo_meta, bams, bedfile ]
+                        fasta: [ fasta ]
+                }
+            SAMTOOLS_MPILEUP_PILEUPCALLER(
+                ch_mpileup_inputs.bams,
+                ch_mpileup_inputs.fasta,
+            )
+            ch_versions = ch_versions.mix( SAMTOOLS_MPILEUP_PILEUPCALLER.out.versions.first() )
+
+            ch_pileupcaller_input = SAMTOOLS_MPILEUP_PILEUPCALLER.out.mpileup
+                .map {
+                    WorkflowEager.addNewMetaFromAttributes( it, "reference", "reference" , false )
+                }
+                .combine( ch_refs_for_mpileup_pileupcaller, by:0 )
+                .multiMap {
+                    ignore_me, meta, mpileup, ref_meta, fasta, fai, dict, bed, snp ->
+                        // def snpfile = snp != "" ? snp : []
+                        mpileup: [ meta, mpileup ]
+                        snpfile: snp
+                }
+
+            // TODO NOTE: Maybe implement a check that unmerged R2 reads have not been kept and throw a warning for ssDNA libs? See: https://github.com/stschiff/sequenceTools/issues/24
```

Comment:
Still want to do this? If yes, maybe make an issue and remove this comment

Reply:
Made an issue. will remove.